runtime/src/main/java/org/corfudb/runtime/collections/Table.java

```diff
@@ -171,6 +179,32 @@ public void clearAll() {
         corfuTable.clear();
     }
 
+    /**
+     * Allow GC to remove all the in-memory objects associated with the CorfuTable
+     * and make it look like the table is a newly opened one
+     * This is useful for JVMs who do not wish to keep the entire table
+     * around in memory.
+     * @param runtime - the runtime that was used to create this table
+     * @param serializer - the serializer used to materialize table data
+     */
+    public void resetTableData(CorfuRuntime runtime, ISerializer serializer) {
```

Comment:
What if there are parallel calls?  We update `this.corfuTable` and that could cause race condition - for example there is a parallel Table.get() and it could hit NPE.

Reply:
NullPointerException in my opinion is one correct way to handle an unsupported operation like resetTableData() in parallel with normal table access. We could make the CorfuTable parameter volatile but that might penalize the vast majority of operations in favor of an unsupported use case. Please let me know what you think?

Reply:
But NPE is not handled by any Table APIs, right? Looks like this is a new failure model and users need to retry upon such exception (e.g. TableResetException).

Reply:
it could be handled if they just catch generic Exception and retry might just work for them. Even if it does not a failure during this freeTableData() + access race condition is not something we should handle for.
I guess the question would be - is there a very cheap way to handle this race condition without penalizing the correct api users?

Reply:
Obviously this isn't intended and it's a bug. `NullPointerException` doesn't prevent parallel access, if there's an access that happens just before `this.corfuTable = null` then there wouldn't be an NPE thrown, yet concurrent access to the old object will be permitted.

Reply:
To be safe, I removed that unnecessary null pointer assignment. Since Java reference variable reads and writes are atomic
the caller should only see one of two values - the old CorfuTable which is ok, or the new one which syncs upon first access. 
https://docs.oracle.com/javase/tutorial/essential/concurrency/atomic.html

Reply:
The issue here is not atomicity, it is consistency. There is still an issue: after creating a new table a concurrent table can still observe the old table reference, although that might not be an issue (not clear), but its still a race.

Reply:
Sorry I missed your last comment, @Maithem - to be sure I just ran a simple test that created 2 instances of CorfuTable and interleaved their accesses to simulate the inconsistency state that would occur if we don't use the `volatile` and so far there appear to be no issues. Let me know if you still see any concerns.